## 1. Data Exploration

### 1.1. Import Libraries

In [38]:
import numpy as np
import pandas as pd
import re

### 1.2. Data Collection

In [39]:
# Read data from file
ds_survey_df = pd.read_csv('../Dataset/kaggle_survey_2022_responses.csv', )
ds_survey_df.head()

C:\Users\WIN 10\AppData\Local\Temp\ipykernel_22372\2297636855.py:2: DtypeWarning: Columns (0,208,225,255,257,260,270,271,277) have mixed types. Specify dtype option on import or set low_memory=False.
  ds_survey_df = pd.read_csv('../Dataset/kaggle_survey_2022_responses.csv', )


Duration (in seconds)                           Q2  \
0  Duration (in seconds)  What is your age (# years)?   
1                    121                        30-34   
2                    462                        30-34   
3                    293                        18-21   
4                    851                        55-59   

                                       Q3  \
0  What is your gender? - Selected Choice   
1                                     Man   
2                                     Man   
3                                     Man   
4                                     Man   

                                          Q4  \
0  In which country do you currently reside?   
1                                      India   
2                                    Algeria   
3                                      Egypt   
4                                     France   

                                                  Q5  \
0  Are you currently a student? (high school, uni...   
1                                                 No   
2                                                 No   
3                                                Yes   
4                                                 No   

                                                Q6_1  \
0  On which platforms have you begun or completed...   
1                                                NaN   
2                                                NaN   
3                                           Coursera   
4                                           Coursera   

                                                Q6_2  \
0  On which platforms have you begun or completed...   
1                                                NaN   
2                                                NaN   
3                                                edX   
4                                                NaN   

                                                Q6_3  \
0  On which platforms have you begun or completed...   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                               Kaggle Learn Courses   

                                                Q6_4  \
0  On which platforms have you begun or completed...   
1                                                NaN   
2                                                NaN   
3                                           DataCamp   
4                                                NaN   

                                                Q6_5  ...  \
0  On which platforms have you begun or completed...  ...   
1                                                NaN  ...   
2                                                NaN  ...   
3                                                NaN  ...   
4                                                NaN  ...   

                                               Q44_3  \
0  Who/what are your favorite media sources that ...   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                                               Q44_4  \
0  Who/what are your favorite media sources that ...   
1                                                NaN   
2                                                NaN   
3                    Kaggle (notebooks, forums, etc)   
4                    Kaggle (notebooks, forums, etc)   

                                               Q44_5  \
0  Who/what are your favorite media sources that ...   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4  Course Forums (forums.fast.ai, Coursera forums...   

                                     

### 1.3. How many rows/columns are there in the dataset?

Before counting rows/columns we should drop the first row because it is just the sub-text for the meaning for each column

In [40]:
ds_survey_df.drop(0, inplace = True) 
ds_survey_df.reset_index(drop= True, inplace = True)
ds_survey_df.head()

Duration (in seconds)     Q2   Q3       Q4   Q5      Q6_1 Q6_2  \
0                   121  30-34  Man    India   No       NaN  NaN   
1                   462  30-34  Man  Algeria   No       NaN  NaN   
2                   293  18-21  Man    Egypt  Yes  Coursera  edX   
3                   851  55-59  Man   France   No  Coursera  NaN   
4                   232  45-49  Man    India  Yes       NaN  NaN   

                   Q6_3      Q6_4 Q6_5  ... Q44_3  \
0                   NaN       NaN  NaN  ...   NaN   
1                   NaN       NaN  NaN  ...   NaN   
2                   NaN  DataCamp  NaN  ...   NaN   
3  Kaggle Learn Courses       NaN  NaN  ...   NaN   
4                   NaN       NaN  NaN  ...   NaN   

                             Q44_4  \
0                              NaN   
1                              NaN   
2  Kaggle (notebooks, forums, etc)   
3  Kaggle (notebooks, forums, etc)   
4                              NaN   

                                               Q44_5  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3  Course Forums (forums.fast.ai, Coursera forums...   
4                                                NaN   

                                               Q44_6  \
0                                                NaN   
1                                                NaN   
2  YouTube (Kaggle YouTube, Cloud AI Adventures, ...   
3                                                NaN   
4                                                NaN   

                                               Q44_7  \
0                                                NaN   
1                                                NaN   
2  Podcasts (Chai Time Data Science, O’Reilly Dat...   
3                                                NaN   
4                                                NaN   

                                               Q44_8 Q44_9 Q44_10 Q44_11  \
0                                                NaN   NaN    NaN    NaN   
1                                                NaN   NaN    NaN    NaN   
2                                                NaN   NaN    NaN    NaN   
3  Blogs (Towards Data Science, Analytics Vidhya,...   NaN    NaN    NaN   
4  Blogs (Towards Data Science, Analytics Vidhya,...   NaN    NaN    NaN   

  Q44_12  
0    NaN  
1    NaN  
2    NaN  
3    NaN  
4    NaN  

[5 rows x 296 columns]

In [41]:
n_rows, n_cols = ds_survey_df.shape

print('Number of rows: ', n_rows)
print('Number of columns: ', n_cols)
print('Columns: ', ds_survey_df.columns)

Number of rows:  23997
Number of columns:  296
Columns:  Index(['Duration (in seconds)', 'Q2', 'Q3', 'Q4', 'Q5', 'Q6_1', 'Q6_2', 'Q6_3',
       'Q6_4', 'Q6_5',
       ...
       'Q44_3', 'Q44_4', 'Q44_5', 'Q44_6', 'Q44_7', 'Q44_8', 'Q44_9', 'Q44_10',
       'Q44_11', 'Q44_12'],
      dtype='object', length=296)


Notice that all columns has data types of object. This makes sense, since all the questions answer are of type string. As for Duration, the suitable data type should be float. 

The details meaning of the columns are aviailable as Kaggle had published them along with the survey information.

Note: "Person who took the survey" wiil be shortened to PWTS

### Details meaning of each questions/columns sorted by theme:
- Durations: The duration the survey takes to complete by a person. (Calculated by system)
### Bibography:
- Q2: Age of PWTS (One choice)
- Q3: Gender of PWTS (One choice)
- Q4: Country that PWTS comes from (Once choice)
- Q5: Whether PWTS a student (One choice)
### Education:
- Q6: Platforms that PWTS have used to study Data Sciecne (Multiple choice)
- Q7: Platforms/Prodects that PWTS found to be most helpful when started studying DS (Multiple choice)
- Q8: Highest level of education that PWTS had attained/plan to attain in the next 2 years (One choice)
- Q9: Whether PWTS had published an academic research (One choice)
- Q10: If PWTS had published an academic research, did the research make use of machine learning? (Multiple choice)
### Programming skills & experience:
- Q11: Number of years the PWTS have been writing code (One choice)
- Q12: Programming languages that PWTS use on a regular basis? (Multiple choice)
- Q13: IDE that PWTS use on a regular basis? (Multiple choice)
- Q14: Notebook hosting products that PWTS use? (Multiple choice)
- Q15: Data visualization libraries that PWTS use on a regular basis? (Multiple choice)
- Q16: Number of years PWTS have used ML methods? (One choice)
- Q17: ML frameworks that PWTS use on a regular basis? (Multiple choice)
- Q18: ML algorithm that PWTS use on a regular basis? (Multiple choice)
- Q19: Computer vision methods that PWTS use on a regular basis? (Multiple choice)
- Q20: NLP (natural language processing) methods that PWTS use on a regular basis? (Multiple choice)
- Q21: Pre-trained model weights services that PWTS download? (Multiple choice)
- Q22: ML model hubs/repositories that PWTS use often? (One choice)
### Employments:
- Q23: Tittle that most similar to PWTS current role? (One choice)
- Q24: Industry that PWTS currently employed/contracted in? (One choice)
- Q25: Size of company that PWTS employed in? (One choice)
- Q26: The approximate number of individuals responsible for DS workloads at PWTS place of work? (One choice)
- Q27: Whether current employer incorporate ML methods in their business? (One choice)
- Q28: Activities that makes up important part of PWTS at work? (Multiple choice)
- Q29: Current yearly compensation of PWTS? (One choice)
### Other product used (Cloud computing, Data storage...):
- Q30: The approximate number of money that PWTS had spent of ML/cloud computer at home/at work in the past 5 years in USD? (One choice)
- Q31: Cloud computing platforms that PWTS uses? (Multiple choice)
- Q32: Cloud platforms that PWTS had most familiar with/has the best developer experience? (One choice)
- Q33: Cloud computing products that PWTS uses? (Multiple choice)
- Q34: Data storage products that PWTS uses? (Multiple choice)
- Q35: Data products (relational databases, data warehouses, data lakes...) that PWTS are most farmiliar with? (Multiple choice)
- Q36: Business Intelligence tools that PWTS uses? (Multiple choice)
- Q37: Managed machine learning products that PWTS uses? (Multiple choice)
- Q38: Automated Machine Learning tools that PWTS uses? (Multiple choice)
- Q39: Products that PWTS to help with ML models? (Multiple choice)
- Q40: Tools that help PWTS monitor ML models/experiments? (Multiple choice)
- Q41: Responsible/Ethical AI products that PWTS uses in ML practices? (Multiple choice)
- Q42: Specilized hardware that PWTS uses when training ML models? (Multiple choice)
- Q43: Times PWTS used TPU (tensor processing unit)? (One choice)
### Source of information:
- Q44: Favorite DS media sources of PWTS? (Multiple choice)

In total, the survey have 44 questions with number of restrictions related to the choice in some questions. For example, if the answer to Q5 answer is that PWTS's a student, questions related to employment would not be asked.

The reason the number of columns in the dataset (296) is not equal to the number of questions in the survey is that each columns represent an option/a choice in every questions.

###  Rename The Columns
- From the above descriptions, we can easily see that the column names are quite long and have many unnecessary words. Therefore, we need to give each column a new name that is shorter and easier to understand.
- The columns can be grouped into two categories: single-choice questions and multiple-choice questions. We will rename each type of column sequentially.

In [42]:
# single-choice questions
one_choice_cols = {
    'Q2': 'Age',
    'Q3': 'Gender',
    'Q4': 'Residential country',
    'Q5': 'Student?', 
    'Q8': 'Higher education?',
    'Q9': 'Academic research publication?',
    'Q11': 'Programming experience',
    'Q16': 'ML experience',
    'Q22': 'Most used ML hub/repository',
    'Q23' : 'Current role',
    'Q24': 'Current industry',
    'Q25' : 'Company size',
    'Q26': 'Individual responsibility',
    'Q27': 'Incorporate ML methods',
    'Q29': 'Current income',
    'Q30' : 'Money spend for ML/Cloud',
    'Q32': 'Familiar Cloud Platform',
    'Q43' : 'TPU experience'
}

In [43]:
# multiple-choice questions
Q6 = {
    'Q6_1': 'DS Platform (Coursera)',
    'Q6_2': 'DS Platform (edX)',
    'Q6_3': 'DS Platform (Kaggle Learn Courses)',
    'Q6_4': 'DS Platform (DataCamp)',
    'Q6_5': 'DS Platform (Fast.ai)',
    'Q6_6': 'DS Platform (Udacity)',
    'Q6_7': 'DS Platform (Udemy)',
    'Q6_8': 'DS Platform (LinkedIn Learning)',
    'Q6_9': 'DS Platform (Cloud-certification programs)',
    'Q6_10': 'DS Platform (University Courses)',
    'Q6_11': 'DS Platform (None)',
    'Q6_12': 'DS Platform (Other)'
}

Q7 = {
    'Q7_1': 'Most helpful Platform (University courses)',
    'Q7_2': 'Most helpful Platform (Online courses)',
    'Q7_3': 'Most helpful Platform (Social media platforms)',
    'Q7_4': 'Most helpful Platform (Video platforms)',
    'Q7_5': 'Most helpful Platform (Kaggle)',
    'Q7_6': 'Most helpful Platform (None / I do not study data science)',
    'Q7_7': 'Most helpful Platform (Other)',
}

Q10 = {
    'Q10_1': 'Use ML (Theoretical research)',
    'Q10_2': 'Use ML (Applied research)',
    'Q10_3': 'Use ML (No)',    
}

Q12 = {
    'Q12_1': 'Programming language (Python)',
    'Q12_2': 'Programming language (R)',
    'Q12_3': 'Programming language (SQL)',
    'Q12_4': 'Programming language (C)',
    'Q12_5': 'Programming language (C#)',
    'Q12_6': 'Programming language (C++)',
    'Q12_7': 'Programming language (Java)',
    'Q12_8': 'Programming language (JavaScript)',
    'Q12_9': 'Programming language (Bash)',
    'Q12_10': 'Programming language (PHP)',
    'Q12_11': 'Programming language (MATLAB)',
    'Q12_12': 'Programming language (Julia)',
    'Q12_13': 'Programming language (Go)',
    'Q12_14': 'Programming language (None)',
    'Q12_15': 'Programming language (Other)',
    
}

Q13 = {
    'Q13_1': 'IDE (JupyterLab)',
    'Q13_2': 'IDE (RStudio/Posit)',
    'Q13_3': 'IDE (Visual Studio)',
    'Q13_4': 'IDE (Visual Studio Code)',
    'Q13_5': 'IDE (PyCharm)',
    'Q13_6': 'IDE (Spyder)',
    'Q13_7': 'IDE (Notepad++)',
    'Q13_8': 'IDE (Sublime Text)',
    'Q13_9': 'IDE (Vim, Emacs)',
    'Q13_10': 'IDE (MATLAB)',
    'Q13_11': 'IDE (Jupyter  Notebook)',
    'Q13_12': 'IDE (IntelliJ)',
    'Q13_13': 'IDE (None)',
    'Q13_14': 'IDE (Other)',
}

Q14 = {
    'Q14_1': 'Hosted notebook product (Kaggle Notebooks)',
    'Q14_2': 'Hosted notebook product (Colab Notebooks)',
    'Q14_3': 'Hosted notebook product (Azure Notebooks)',
    'Q14_4': 'Hosted notebook product (Code Ocean)',
    'Q14_5': 'Hosted notebook product (IBM Watson Studio)',
    'Q14_6': 'Hosted notebook product (Amazon Sagemaker Studio)',
    'Q14_7': 'Hosted notebook product (Amazon Sagemaker Studio Lab)',
    'Q14_8': 'Hosted notebook product (Amazon EMR Notebooks )',
    'Q14_9': 'Hosted notebook product (Google Cloud Vertex AI Workbench)',
    'Q14_10': 'Hosted notebook product (Hex Workspaces)',
    'Q14_11': 'Hosted notebook product (Noteable Notebooks)',
    'Q14_12': 'Hosted notebook product (Databricks Collaborative Notebooks)',
    'Q14_13': 'Hosted notebook product (Deepnote Notebooks)',
    'Q14_14': 'Hosted notebook product (Gradient Notebooks)',
    'Q14_15': 'Hosted notebook product (None)',
    'Q14_16': 'Hosted notebook product (Other)',
}

Q15 = {
    'Q15_1' : 'Data visualization libraries (Matplotlib)',
    'Q15_2' : 'Data visualization libraries (Seaborn)',
    'Q15_3' : 'Data visualization libraries (Plotly/ Plotky Express)',
    'Q15_4' : 'Data visualization libraries (Ggplot/ ggplot2)',
    'Q15_5' : 'Data visualization libraries (Shiny)',
    'Q15_6' : 'Data visualization libraries (D3 js)',
    'Q15_7' : 'Data visualization libraries (Altair)',
    'Q15_8' : 'Data visualization libraries (Bokeh)',
    'Q15_9' : 'Data visualization libraries (Geoplotlib)',
    'Q15_10' : 'Data visualization libraries (Leaflet/ Folium)',
    'Q15_11' : 'Data visualization libraries (Pygal)',
    'Q15_12' : 'Data visualization libraries (Dygraphs)',
    'Q15_13' : 'Data visualization libraries (Highcharter)',
    'Q15_14' : 'Data visualization libraries (None)',
    'Q15_15' : 'Data visualization libraries (Other)',
}

Q17 = {
    'Q17_1' : 'ML frameworks (Scikit-learn)',
    'Q17_2' : 'ML frameworks (TensorFlow)',
    'Q17_3' : 'ML frameworks (Keras)',
    'Q17_4' : 'ML frameworks (Pytorch)',
    'Q17_5' : 'ML frameworks (Fast.ai)',
    'Q17_6' : 'ML frameworks (Xgboost)',
    'Q17_7' : 'ML frameworks (LightGBM)',
    'Q17_8' : 'ML frameworks (CatBoost)',
    'Q17_9' : 'ML frameworks (Caret)',
    'Q17_10' : 'ML frameworks (Tidymodels)',
    'Q17_11' : 'ML frameworks (JAX)',
    'Q17_12' : 'ML frameworks (Pytorch Lightning)',
    'Q17_13' : 'ML frameworks (Huggingface)',
    'Q17_14' : 'ML frameworks (None)',
    'Q17_15' : 'ML frameworks (Other)',
}

Q18 = {
    'Q18_1': 'ML Algorithms (Linear or Logistict Regression)',
    'Q18_2': 'ML Algorithms (Decison Trees or Random Forests)',
    'Q18_3': 'ML Algorithms (Gradient Boosting Machines)',
    'Q18_4': 'ML Algorithms (Bayesian Approaches)',
    'Q18_5': 'ML Algorithms (Evolutionary Approaches)',
    'Q18_6': 'ML Algorithms (Dense Neural Networks)',
    'Q18_7': 'ML Algorithms (Convolational Neural Networks)',
    'Q18_8': 'ML Algorithms (Generative Adverarial Networks)',
    'Q18_9': 'ML Algorithms (Recurrent Neural Networks)',
    'Q18_10': 'ML Algorithms (Transformer Networks)',
    'Q18_11': 'ML Algorithms (Autoencoder Networks)',
    'Q18_12': 'ML Algorithms (Graph Neural Networks)',
    'Q18_13': 'ML Algorithms (None)',
    'Q18_14': 'ML Algorithms (Other)',
}

Q19 = {
    'Q19_1' : 'Computer visions method (General tools)',
    'Q19_2' : 'Computer visions method (Image segmentation methods)',
    'Q19_3' : 'Computer visions method (Object detection methods)',
    'Q19_4' : 'Computer visions method (Image classification)',
    'Q19_5' : 'Computer visions method (Generative networks)',
    'Q19_6' : 'Computer visions method (Vision transformer networks)',
    'Q19_7' : 'Computer visions method (None)',
    'Q19_8' : 'Computer visions method (Other)',
}

Q20 = {
    'Q20_1' : 'NLP methods (Word embeddings/vectors)',
    'Q20_2' : 'NLP methods (Encoder-decoder models)',
    'Q20_3' : 'NLP methods (Contextualized embeddings)',
    'Q20_4' : 'NLP methods (Transformer language models)',
    'Q20_5' : 'NLP methods (None)',
    'Q20_6' : 'NLP methods (Other)',
}

Q21 = {
    'Q21_1' : 'Pretrain model services (TensorFlow Hub)',
    'Q21_2' : 'Pretrain model services (Pytorch Hub)',
    'Q21_3' : 'Pretrain model services (Huggingface models)',
    'Q21_4' : 'Pretrain model services (Timm)',
    'Q21_5' : 'Pretrain model services (Jumpstart)',
    'Q21_6' : 'Pretrain model services (ONNX models)',
    'Q21_7' : 'Pretrain model services (NVIDA NGC models)',
    'Q21_8' : 'Pretrain model services (Kaggle datasets)',
    'Q21_9' : 'Pretrain model services (None)',
    'Q21_10' : 'Pretrain model services (Other)',
}

Q28 = {
    'Q28_1' : 'Important role at work (Analyze and understand data)',
    'Q28_2' : 'Important role at work (Build/ run data infrastructure)',
    'Q28_3' : 'Important role at work (Build ML prototypes)',
    'Q28_4' : 'Important role at work ( Build/run ML services)',
    'Q28_5' : 'Important role at work (Improve existing ML models)',
    'Q28_6' : 'Important role at work (Research, advance ML models)',
    'Q28_7' : 'Important role at work (None)',
    'Q28_8' : 'Important role at work (Other)',
}

Q31 = {
    'Q31_1' : 'Cloud computing Platforms (Amazon Web Services)',
    'Q31_2' : 'Cloud computing Platforms (Microsoft Azure)',
    'Q31_3' : 'Cloud computing Platforms (Google Cloud Platform)',
    'Q31_4' : 'Cloud computing Platforms (IBM Cloud/ Red Hat)',
    'Q31_5' : 'Cloud computing Platforms (Oracle Cloud)',
    'Q31_6' : 'Cloud computing Platforms (SAP Cloud)',
    'Q31_7' : 'Cloud computing Platforms (VMware Cloud)',
    'Q31_8' : 'Cloud computing Platforms (Alibaba Cloud)',
    'Q31_9' : 'Cloud computing Platforms (Tencent Cloud)',
    'Q31_10' : 'Cloud computing Platforms (Huawei Cloud)',
    'Q31_11' : 'Cloud computing Platforms (None)',
    'Q31_12' : 'Cloud computing Platforms (Other)',
}

Q33 = {
    'Q33_1' : 'Cloud Computing products use (Amazon Elastic Compute Coud)',
    'Q33_2' : 'Cloud Computing products use (Microsoft Axure Virtual Machines)',
    'Q33_3' : 'Cloud Computing products use (Google Cloud Compute Engine)',
    'Q33_4' : 'Cloud Computing products use (None)',
    'Q33_5' : 'Cloud Computing products use (Other)',
}

Q34 = {
    'Q34_1' : 'Data storage products (Microsoft Azure Blob Storage)',
    'Q34_2' : 'Data storage products (Microsoft Azure Files)',
    'Q34_3' : 'Data storage products (Amazon Simple Storage Service)',
    'Q34_4' : 'Data storage products (Amazon Elastic File System)',
    'Q34_5' : 'Data storage products (Google CLoud Storage)',
    'Q34_6' : 'Data storage products (Gooogle Cloud Filestore)',
    'Q34_7' : 'Data storage products (None)',
    'Q34_8' : 'Data storage products (Other)',
}

Q35 = {
    'Q35_1' : 'Data products (MySQL)',
    'Q35_2' : 'Data products (PosgreSQL)',
    'Q35_3' : 'Data products (SQLite)',
    'Q35_4' : 'Data products (Oracle Database)',
    'Q35_5' : 'Data products (MongoDB)',
    'Q35_6' : 'Data products (SnowFlake)',
    'Q35_7' : 'Data products (IBM Db2)',
    'Q35_8' : 'Data products (Microsoft SQL Server)',
    'Q35_9' : 'Data products (Micorsoft Azure SQL Database)',
    'Q35_10' : 'Data products (Amazon Redshift)',
    'Q35_11' : 'Data products (Amazon RDS)',
    'Q35_12' : 'Data products (Amazon DynamoDB)',
    'Q35_13' : 'Data products (Google Cloud BigQuery)',
    'Q35_14' : 'Data products (Google Cloud SQL)',
    'Q35_15' : 'Data products (None)',
    'Q35_16' : 'Data products (Other)',
}

Q36 = {
    'Q36_1' : 'Business Intelligent tools (Amazon QuickSight)',
    'Q36_2' : 'Business Intelligent tools (Microsoft Power BI)',
    'Q36_3' : 'Business Intelligent tools (Google Data Studio)',
    'Q36_4' : 'Business Intelligent tools (Looker)',
    'Q36_5' : 'Business Intelligent tools (Tablue)',
    'Q36_6' : 'Business Intelligent tools (Qlik Sense)',
    'Q36_7' : 'Business Intelligent tools (Domo)',
    'Q36_8' : 'Business Intelligent tools (TIBCO Spotire)',
    'Q36_9' : 'Business Intelligent tools (ALteryx)',
    'Q36_10' : 'Business Intelligent tools (Silense)',
    'Q36_11' : 'Business Intelligent tools (SAP Ana;ytics Cloud)',
    'Q36_12' : 'Business Intelligent tools (Microsoft Azure Sysnapse)',
    'Q36_13' : 'Business Intelligent tools (Thoughtspot)',
    'Q36_14' : 'Business Intelligent tools (None)',
    'Q36_15' : 'Business Intelligent tools (Other)',
}

Q37 = {
    'Q37_1' : 'ML products (Amazon SageMaker)',
    'Q37_2' : 'ML products (Azure Machine Learning Studio)',
    'Q37_3' : 'ML products (Google CLoud Vertex AI)',
    'Q37_4' : 'ML products (DataRobot)',
    'Q37_5' : 'ML products (Databricks)',
    'Q37_6' : 'ML products (Dataiku)',
    'Q37_7' : 'ML products (Alteryx)',
    'Q37_8' : 'ML products (Rapidminer)',
    'Q37_9' : 'ML products (C3.ai)',
    'Q37_10' : 'ML products (Domino Data Lab)',
    'Q37_11' : 'ML products (H2O AI Cloud)',
    'Q37_12' : 'ML products (None)',
    'Q37_13' : 'ML products (Other)',

}

Q38 = {
    'Q38_1' : 'Automated ML tools (Google Cloud AutoML)',
    'Q38_2' : 'Automated ML tools (H2O Driverless AI)',
    'Q38_3' : 'Automated ML tools (Databricks AutoML)',
    'Q38_4' : 'Automated ML tools (DataRobot AutoML)',
    'Q38_5' : 'Automated ML tools (Amazon Sagemaker Autopilot)',
    'Q38_6' : 'Automated ML tools (Azure Automated Machine Learning)',
    'Q38_7' : 'Automated ML tools (None)',
    'Q38_8' : 'Automated ML tools (Other)',
}

Q39 = {
    'Q39_1' : 'ML products use (TensorFlow Extendted)',
    'Q39_2' : 'ML products use (TorchServe)',
    'Q39_3' : 'ML products use (ONNX Runtime)',
    'Q39_4' : 'ML products use (Triton Inference Server)',
    'Q39_5' : 'ML products use (Open VINO Model Server)',
    'Q39_6' : 'ML products use (KServe)',
    'Q39_7' : 'ML products use (BentoML)',
    'Q39_8' : 'ML products use (Multi Model Server)',
    'Q39_9' : 'ML products use (Seldon Core)',
    'Q39_10' : 'ML products use (ML flow)',
    'Q39_11' : 'ML products use (None)',
    'Q39_12' : 'ML products use (Other)',
}

Q40 = {
    'Q40_1' : 'ML monitors (Neptune.ai)',
    'Q40_2' : 'ML monitors (Weights & Biases)',
    'Q40_3' : 'ML monitors (Comet.ml)',
    'Q40_4' : 'ML monitors (TensorBoard)',
    'Q40_5' : 'ML monitors (Guild.ai)',
    'Q40_6' : 'ML monitors (ClearML)',
    'Q40_7' : 'ML monitors (MLflow)',
    'Q40_8' : 'ML monitors (Aporia)',
    'Q40_9' : 'ML monitors (Evidently AI)',
    'Q40_10' : 'ML monitors (Arize)',
    'Q40_11' : 'ML monitors (WhyLabs)',
    'Q40_12' : 'ML monitors (Fiddler)',
    'Q40_13' : 'ML monitors (DVC)',
    'Q40_14' : 'ML monitors (None)',
    'Q40_15' : 'ML monitors (Other)',
}

Q41 = {
    'Q41_1' : 'AI products (Google)',
    'Q41_2' : 'AI products (Microsoft)',
    'Q41_3' : 'AI products (IBM)',
    'Q41_4' : 'AI products (Amazon)',
    'Q41_5' : 'AI products (The LinkedIn)',
    'Q41_6' : 'AI products (Audit-AI)',
    'Q41_7' : 'AI products (Aequitas)',
    'Q41_8' : 'AI products (None)',
    'Q41_9' : 'AI products (Other)',
}

Q42 = {
    'Q42_1' : 'Hardware use (GPUs)',
    'Q42_2' : 'Hardware use (TPUs)',
    'Q42_3' : 'Hardware use (IPUs)',
    'Q42_4' : 'Hardware use (RDUs)',
    'Q42_5' : 'Hardware use (WSEs)',
    'Q42_6' : 'Hardware use (Trainium Chips)',
    'Q42_7' : 'Hardware use (Inferentia Chips)',
    'Q42_8' : 'Hardware use (None)',
    'Q42_9' : 'Hardware use (Other)',
}

Q44 = {
    'Q44_1' : 'Favorite media sources (Twitter)',
    'Q44_2' : 'Favorite media sources (Email newsletters)',
    'Q44_3' : 'Favorite media sources (Reddit)',
    'Q44_4' : 'Favorite media sources (Kaggle)',
    'Q44_5' : 'Favorite media sources (Course Forums)',
    'Q44_6' : 'Favorite media sources (YouTube)',
    'Q44_7' : 'Favorite media sources (Podcasts)',
    'Q44_8' : 'Favorite media sources (Blogs)',
    'Q44_9' : 'Favorite media sources (Journal Publications)',
    'Q44_10' : 'Favorite media sources (Slack)',
    'Q44_11' : 'Favorite media sources (None)',
    'Q44_12' : 'Favorite media sources (Other)',

}

In [44]:
# Replace columns names
new_cols_name = [one_choice_cols, Q6, Q7, Q10, Q12, 
                 Q13, Q14, Q15, Q17, Q18, Q19, Q20, 
                 Q21, Q28, Q31, Q33, Q34, Q35, 
                 Q36, Q37, Q38, Q39, Q40, Q41, Q42, Q44]

# Flatten the list of dictionaries into a single dictionary
new_cols_dict = {k: v for d in new_cols_name for k, v in d.items()}

# Rename columns in the DataFrame
ds_survey_df.rename(columns=new_cols_dict, inplace=True)

Let's check our new `ds_survey_df`

In [45]:
ds_survey_df

Duration (in seconds)    Age Gender       Residential country Student?  \
0                       121  30-34    Man                     India       No   
1                       462  30-34    Man                   Algeria       No   
2                       293  18-21    Man                     Egypt      Yes   
3                       851  55-59    Man                    France       No   
4                       232  45-49    Man                     India      Yes   
...                     ...    ...    ...                       ...      ...   
23992                   331  22-24    Man  United States of America      Yes   
23993                   330  60-69    Man  United States of America      Yes   
23994                   860  25-29    Man                    Turkey       No   
23995                   597  35-39  Woman                    Israel       No   
23996                   303  18-21    Man                     India      Yes   

      DS Platform (Coursera) DS Platform (edX)  \
0                        NaN               NaN   
1                        NaN               NaN   
2                   Coursera               edX   
3                   Coursera               NaN   
4                        NaN               NaN   
...                      ...               ...   
23992                    NaN               NaN   
23993                    NaN               NaN   
23994                    NaN               NaN   
23995                    NaN               NaN   
23996                    NaN               NaN   

      DS Platform (Kaggle Learn Courses) DS Platform (DataCamp)  \
0                                    NaN                    NaN   
1                                    NaN                    NaN   
2                                    NaN               DataCamp   
3                   Kaggle Learn Courses                    NaN   
4                                    NaN                    NaN   
...                                  ...                    ...   
23992                                NaN                    NaN   
23993                                NaN                    NaN   
23994                                NaN               DataCamp   
23995               Kaggle Learn Courses                    NaN   
23996                                NaN                    NaN   

      DS Platform (Fast.ai)  ... Favorite media sources (Reddit)  \
0                       NaN  ...                             NaN   
1                       NaN  ...                             NaN   
2                       NaN  ...                             NaN   
3                       NaN  ...                             NaN   
4                       NaN  ...                             NaN   
...                     ...  ...                             ...   
23992                   NaN  ...                             NaN   
23993                   NaN  ...                             NaN   
23994                   NaN  ...                             NaN   
23995                   NaN  ...                             NaN   
23996                   NaN  ...                             NaN   

       Favorite media sources (Kaggle)  \
0                                  NaN   
1                                  NaN   
2      Kaggle (notebooks, forums, etc)   
3      Kaggle (notebooks, forums, etc)   
4                                  NaN   
...                                ...   
23992  Kaggle (notebooks, forums, etc)   
23993                              NaN   
23994  Kaggle (notebooks, forums, etc)   
23995                              NaN   
23996                              NaN   

                  Favorite media sources (Course Forums)  \
0                                                    NaN   
1                                                    NaN   
2                                                    NaN   
3      Course Forums (forums.fast.ai, Coursera forums...   
4                

As you see, we can observe lots of missing values here.

### Meaning Of Rows
Each row is an response to the Kaggle state of Data Science Survey - 2022.

Let's check if there's any duplicated rows

In [46]:
# Check for duplicated rows
duplicated_rows = ds_survey_df[ds_survey_df.duplicated()]

num_duplicated_rows = len(duplicated_rows)
print(f"Number of duplicated rows: {num_duplicated_rows}")

Number of duplicated rows: 0


Great! We have no duplicated rows.

### 1.4. Check The Inappropriate Data Types

In [47]:
ds_survey_df.dtypes

Duration (in seconds)                            object
Age                                              object
Gender                                           object
Residential country                              object
Student?                                         object
                                                  ...  
Favorite media sources (Blogs)                   object
Favorite media sources (Journal Publications)    object
Favorite media sources (Slack)                   object
Favorite media sources (None)                    object
Favorite media sources (Other)                   object
Length: 296, dtype: object

Like we've said above, as for Duration, the suitable data type should be float.

In [48]:
ds_survey_df['Duration (in seconds)'] = ds_survey_df['Duration (in seconds)'].astype(np.float64)

In [49]:
ds_survey_df.dtypes

Duration (in seconds)                            float64
Age                                               object
Gender                                            object
Residential country                               object
Student?                                          object
                                                  ...   
Favorite media sources (Blogs)                    object
Favorite media sources (Journal Publications)     object
Favorite media sources (Slack)                    object
Favorite media sources (None)                     object
Favorite media sources (Other)                    object
Length: 296, dtype: object

### 1.5. With Each Numerical Column, How Are Values Distributed?

In [50]:
def missing_ratio(column):
    missing_values = column.isnull().sum()
    total_values = len(column)
    return (missing_values / total_values) * 100

In [51]:
# Select numeric columns
num_col_info_df = ds_survey_df.select_dtypes(include=['float64']).dropna(axis=1)

missing_ratios = num_col_info_df.agg([missing_ratio])
attribute = num_col_info_df.describe()

num_col_info_df = pd.concat([missing_ratios, attribute], axis = 0)
num_col_info_df

Duration (in seconds)
missing_ratio           0.000000e+00
count                   2.399700e+04
mean                    1.009010e+04
std                     1.115403e+05
min                     1.200000e+02
25%                     2.640000e+02
50%                     4.140000e+02
75%                     7.150000e+02
max                     2.533678e+06

In this dataset, there is only the Duration (in seconds) column is numeric columns.
The maximum duration takes to complete the survey is very strange.

### 1.6. With Each Categorical Column, How Are Values Distributed?

#### 1.6.1. Single-choice Questions
First let's check ratio of missing values:

In [52]:
one_choice_cols = ['Age', 'Gender',  'Residential country', 'Student?', 'Higher education?', 'Academic research publication?',
                   'Programming experience', 'ML experience', 'Current role', 'Current industry', 'Company size','Individual responsibility',
                    'Incorporate ML methods', 'Current income', 'Money spend for ML/Cloud', 'Familiar Cloud Platform', 'TPU experience']
one_choice_cols_df = ds_survey_df[one_choice_cols]

missing_ratios_df = one_choice_cols_df.agg(missing_ratio).to_frame()
missing_ratios_df.columns = ['Missing ratio']

missing_ratios_df

Missing ratio
Age                                  0.000000
Gender                               0.000000
Residential country                  0.000000
Student?                             0.000000
Higher education?                    2.496145
Academic research publication?      48.489395
Programming experience               3.142059
ML experience                       17.131308
Current role                        55.702796
Current industry                    62.103596
Company size                        62.220278
Individual responsibility           62.536984
Incorporate ML methods              63.028712
Current income                      66.095762
Money spend for ML/Cloud            64.645581
Familiar Cloud Platform             91.961495
TPU experience                      80.860108

Here are some observations based on the missing ratios dataframe:

- `Age`, `Gender`, `Residential country`, and `Student?` columns have no missing values (0% missing ratio).

- `Higher education?` and `Programming experience` column has a missing ratio of 2.5% and 3.1%, indicating that a small percentage of survey participants did not provide information about their higher education and progamming experience.

- All other columns, also have the average missing ratios.

- Specially, `Familiar Cloud Platform` and `TPU experience` columns have a very large precentage of missing values (up to 92% and 80%).
As we can see the ratio is better adjusted for non-student dataset. Still, there are some fields with a lot of missing values (> 50% is missing) - `Familiar Cloud Platform` and `TPU experience`. `The Kaggle survey 2022 answer choices` state that the 2 fields is asked after another fields related is answered which is why they have so many missing values.

Next, let's explore the distribution of the 1 choice answer data.

In [53]:
one_choice_response_distribution = pd.DataFrame(columns=['Field', 'Distribution'])
for col in one_choice_cols_df.columns:
    one_choice_response_distribution.loc[len(one_choice_response_distribution)] = [col, one_choice_cols_df[col].value_counts(dropna=False).to_dict()]
one_choice_response_distribution

Field  \
0                              Age   
1                           Gender   
2              Residential country   
3                         Student?   
4                Higher education?   
5   Academic research publication?   
6           Programming experience   
7                    ML experience   
8                     Current role   
9                 Current industry   
10                    Company size   
11       Individual responsibility   
12          Incorporate ML methods   
13                  Current income   
14        Money spend for ML/Cloud   
15         Familiar Cloud Platform   
16                  TPU experience   

                                         Distribution  
0   {'18-21': 4559, '25-29': 4472, '22-24': 4283, ...  
1   {'Man': 18266, 'Woman': 5286, 'Prefer not to s...  
2   {'India': 8792, 'United States of America': 29...  
3                         {'No': 12036, 'Yes': 11961}  
4   {'Master’s degree': 9142, 'Bachelor’s degree':...  
5               {nan: 11636, 'No': 7117, 'Yes': 5244}  
6   {'1-3 years': 6459, '< 1 years': 5454, '3-5 ye...  
7   {'Under 1 year': 7221, nan: 4111, '1-2 years':...  
8   {nan: 13367, 'Data Scientist': 1929, 'Data Ana...  
9   {nan: 14903, 'Computers/Technology': 2321, 'Ac...  
10  {nan: 14931, '0-49 employees': 2130, '10,000 o...  
11  {nan: 15007, '20+': 2270, '1-2': 1815, '0': 14...  
12  {nan: 15125, 'No (we do not use ML methods)': ...  
13  {nan: 15861, '$0-999': 1112, '10,000-14,999': ...  
14  {nan: 15513, '$0 ($USD)': 2822, '$100-$999': 2...  
15  {nan: 22068, ' Amazon Web Services (AWS) ': 55...  
16  {nan: 19404, 'Never': 2538, '2-5 times': 958, ...

#### 1.6.2. Multiple-choice Questions

In [54]:
multi_choice_column = [Q6, Q7, Q10, Q12, 
                 Q13, Q14, Q15, Q17, Q18, Q19, Q20, 
                 Q21, Q28, Q31, Q33, Q34, Q35, 
                 Q36, Q37, Q38, Q39, Q40, Q41, Q42, Q44]

multi_choice_column_distribution = pd.DataFrame(columns=['Field','Distribution'])

for column in multi_choice_column:
    column_distribution = ds_survey_df[column.values()].count()
    column_name = re.search(r'^[^()]+', column_distribution.index[0]).group().strip()
    column_distribution.index = column_distribution.index.str.extract(r'\(([^)]+)\)')[0].str.strip()
    multi_choice_column_distribution.loc[len(multi_choice_column_distribution)] = [column_name, column_distribution.to_dict()]

multi_choice_column_distribution

Field  \
0                    DS Platform   
1          Most helpful Platform   
2                         Use ML   
3           Programming language   
4                            IDE   
5        Hosted notebook product   
6   Data visualization libraries   
7                  ML frameworks   
8                  ML Algorithms   
9        Computer visions method   
10                   NLP methods   
11       Pretrain model services   
12        Important role at work   
13     Cloud computing Platforms   
14  Cloud Computing products use   
15         Data storage products   
16                 Data products   
17    Business Intelligent tools   
18                   ML products   
19            Automated ML tools   
20               ML products use   
21                   ML monitors   
22                   AI products   
23                  Hardware use   
24        Favorite media sources   

                                         Distribution  
0   {'Coursera': 9699, 'edX': 2474, 'Kaggle Learn ...  
1   {'University courses': 6851, 'Online courses':...  
2   {'Theoretical research': 1549, 'Applied resear...  
3   {'Python': 18653, 'R': 4571, 'SQL': 9620, 'C':...  
4   {'JupyterLab': 4887, 'RStudio/Posit': 3824, 'V...  
5   {'Kaggle Notebooks': 7478, 'Colab Notebooks': ...  
6   {'Matplotlib': 14010, 'Seaborn': 10512, 'Plotl...  
7   {'Scikit-learn': 11403, 'TensorFlow': 7953, 'K...  
8   {'Linear or Logistict Regression': 11338, 'Dec...  
9   {'General tools': 2293, 'Image segmentation me...  
10  {'Word embeddings/vectors': 2241, 'Encoder-dec...  
11  {'TensorFlow Hub': 3087, 'Pytorch Hub': 2012, ...  
12  {'Analyze and understand data': 4999, 'Build/ ...  
13  {'Amazon Web Services': 2346, 'Microsoft Azure...  
14  {'Amazon Elastic Compute Coud': 1609, 'Microso...  
15  {'Microsoft Azure Blob Storage': 615, 'Microso...  
16  {'MySQL': 2233, 'PosgreSQL': 1516, 'SQLite': 1...  
17  {'Amazon QuickSight': 224, 'Microsoft Power BI...  
18  {'Amazon SageMaker': 561, 'Azure Machine Learn...  
19  {'Google Cloud AutoML': 463, 'H2O Driverless A...  
20  {'TensorFlow Extendted': 522, 'TorchServe': 16...  
21  {'Neptune.ai': 128, 'Weights & Biases': 381, '...  
22  {'Google': 289, 'Microsoft': 206, 'IBM': 128, ...  
23  {'GPUs': 2682, 'TPUs': 653, 'IPUs': 67, 'RDUs'...  
24  {'Twitter': 3995, 'Email newsletters': 3787, '...

Up to now, we have get some overview information about the dataset. Also rename columns the dataset in a more meaningful way, explored how values in each column is distributed. For a clearly view how values is distributed in each column, we will visualize them in the next part. Save the dataframe

Save the data

In [55]:
ds_survey_df.to_csv('../Dataset/final_data.csv',index= False)